File: backend/chessapp/domain/pieces/pawn.py
from ...domain.pieces.piece import Piece
from ...domain.pieces.piece_type import PieceType
from ...domain.value_objects.side import Side

class Pawn(Piece):

    def get_points(self):
        return 10

    def __init__(self, side: Side):
        super().__init__(side, PieceType.Pawn)
        self._has_moved = False

    def move(self):
        self._has_moved = True

    def has_moved(self):
        return self._has_moved

File: frontend/src/app/pages/game/play/play/models/pieces/pawn.ts
import { Cell } from "../board/ cell";
import { Side } from "../side";
import { Piece } from "./piece";
import { PieceType } from "./piece_type";

export class Pawn extends Piece {

    override get can_move_over(): boolean {
        return false;
    }

    constructor(side: Side) {
        super(side, PieceType.Pawn);
    }

    override validatePush(from: Cell, to: Cell): boolean {
        const [delta_file, delta_rank] = this.calculateMoveDeltas(from, to);

        if (this.side === Side.white) {
            // White pawns move up (+Rank)
            return delta_file === 0 && (delta_rank === 1 || (delta_rank === 2 && from.rank === 2));
        } else {
            // Black pawns move down (-Rank)
            return delta_file === 0 && (delta_rank === -1 || (delta_rank === -2 && from.rank === 7));
        }
    }

    override validateCapture(from: Cell, to: Cell): boolean {
        const [delta_file, delta_rank] = this.calculateMoveDeltas(from, to);

        if (this.side === Side.white) {
            return Math.abs(delta_file) === 1 && delta_rank === 1;
        } else {
            return Math.abs(delta_file) === 1 && delta_rank === -1;
        }
    }
}